File: src/main/java/algo/interval/IntervalListIntersections_986.java
package algo.interval;

import java.util.ArrayList;

public class IntervalListIntersections_986 {
  public static int[][] intervalIntersection(int[][] firstList, int[][] secondList) {
    var intersections = new ArrayList<int[]>();
    int i = 0, j = 0;
    int start, end;
    while (i < firstList.length && j < secondList.length) {
      start = Math.max(firstList[i][0], secondList[j][0]);
      end = Math.min(firstList[i][1], secondList[j][1]);
      if (start <= end) {
        intersections.add(new int[]{start, end});
      }
      if (firstList[i][1] <= secondList[j][1]) {
        i++;
      } else {
        j++;
      }
    }

    var result = new int[intersections.size()][];
    for (i = 0; i < intersections.size(); i++) {
      result[i] = intersections.get(i);
    }

    return result;
  }
}

File: src/main/kotlin/algo/interval/IntervalListIntersections_986.kt
package algo.interval

class KtIntervalListIntersections_986 {
    fun intervalIntersection(firstList: Array<IntArray>, secondList: Array<IntArray>): Array<IntArray> {
        val intersections: MutableList<IntArray> = mutableListOf()
        var i = 0; var j = 0
        while (i < firstList.size && j < secondList.size) {
            val first = firstList[i]
            val second = secondList[j]
            val start = first[0].coerceAtLeast(second[0])
            val end = first[1].coerceAtMost(second[1])
            // there is intersection
            if (start <= end) {
                intersections.add(intArrayOf(start, end))
            }

            if (first[1] <= second[1]) {
                i++
            } else {
                j++
            }
        }
        return intersections.toTypedArray()
    }
}